app/src/gui/components/record/form.tsx

```diff
@@ -987,286 +1017,263 @@ class RecordForm extends React.Component<any, RecordFormState> {
+   * @returns {string} route path to be passed to navigateTo
+   */
+  getRoute(
+    type: 'project' | 'create',
+    params: {
+      recordId?: string;
+      revisionId?: string;
+      formType?: string | null;
+    } = {}
+  ) {
+    const {serverId, project_id} = this.props;
+    const base = ROUTES.INDIVIDUAL_NOTEBOOK_ROUTE + serverId + '/' + project_id;
+
+    switch (type) {
+      case 'project':
+        return base;
+      case 'create':
+        if (!params.formType)
+          throw new Error('Form type required for create route');
+        return base + ROUTES.RECORD_CREATE + params.formType;
+      default:
+        return base;
+    }
+  }
+
+  /**
+   * Handle redirect after a record is saved
+   * - navigates to the appropriate place based on the user selection
+   *   in is_close, passing the right context as needed.
+   *
+   * @param ids - record and revision ids
+   * @param is_close - close or new
+   * @param setSubmitting - function to set the submitting state
+   * @returns revision id of the recently saved child record
+   */
+  redirectAfterSave(
+    ids: RecordIdentifiers,
+    is_close: string,
+    setSubmitting: any
+  ) {
+    const relationState = this.props.location?.state;
+    // first case is if we have a parent record, there should be
+    // some location state passed in
+    if (relationState !== undefined && relationState !== null) {
+      if (
+        recordIsChildRecord(
+          ids.revision_id,
+          relationState,
+          this.props.record_id
+        )
+      )
+        return this.navigateAfterSaveChild(ids, is_close, setSubmitting);
+      else return this.simpleFinishOrNew(ids, is_close, setSubmitting);
+    } else {
+      // here we don't have location state passed in, check the current state
+      // to see if we have recorded a relationship
+      // when is this set?
+      const relationship = this.state.relationship;
+      if (
+        relationship === undefined ||
+        relationship === null ||
+        relationship.parent === undefined ||
+        relationship.parent === null
+      ) {
+        return this.simpleFinishOrNew(ids, is_close, setSubmitting);
+      } else {
+        generateLocationState(
+          relationship.parent,
+          this.props.project_id,
+          this.props.serverId
+        )
+          .then(locationState => {
+            return this.navigateAfterSaveChild(
+              ids,
+              is_close,
+              setSubmitting,
+              locationState.location_state
+            );
+          })
+          .catch(error => logError(error));
+      }
+    }
+  }
+
+  /**
+   * Simple case where we just redirect back to the project page or
+   * a new record creation page
+   *
+   * @param ids - record and revision ids
+   * @param is_close - close or new
+   * @param setSubmitting - function to set the submitting state
+   * @returns revision id of the recently saved child record
+   */
+  simpleFinishOrNew(
```

Comment:
As discussed offline - could you use the object spread arg style rather than ordered anonymous arguments, it's a lot more maintainble/clear.